source/draw/gpu/opengl/OpenGLCanvas.ooc

```diff
@@ -28,6 +28,38 @@ OpenGLCanvas: class extends OpenGLSurface {
 	_target: OpenGLPacked
 	_renderTarget: GLFramebufferObject
 	context ::= this _context as OpenGLContext
+	draw: override func ~DrawState (drawState: DrawState) {
+		// Set shader
+		gpuMap: GpuMap = (drawState map == null) ? this context defaultMap : drawState map as GpuMap
+		// Set viewport
+		viewport := (drawState viewport hasZeroArea) ? IntBox2D new(this size) : drawState viewport
+		this context backend setViewport(viewport)
+		// Set transform
+		gpuMap view = _toLocal * drawState getTransformNormalized() normalizedToReference(this size) * _toLocal
+		// Set model and projection matrices
+		if (this _focalLength > 0.0f) {
+			a := 2.0f * this _focalLength / this size x
+			f := -(this _coordinateTransform e as Float) * 2.0f * this _focalLength / this size y
+			k := (this _farPlane + this _nearPlane) / (this _farPlane - this _nearPlane)
+			o := 2.0f * this _farPlane * this _nearPlane / (this _farPlane - this _nearPlane)
+			gpuMap projection = FloatTransform3D new(a, 0.0f, 0.0f, 0.0f, 0.0f, f, 0.0f, 0.0f, 0.0f, 0.0f, k, -1.0f, 0.0f, 0.0f, o, 0.0f)
+		} else
+			gpuMap projection = FloatTransform3D createScaling(2.0f / this size x, -(this _coordinateTransform e as Float) * 2.0f / this size y, 1.0f)
+		gpuMap model = this _createModelTransform(IntBox2D new(this size))
```

Comment:
This is the current model:
**Model matrix** - the world coordinates of the image plane, i.e. the left top corner of the quad is in (-width / 2, height / 2, focalLength)
**View matrix** - the inverse camera matrix, this is where the stabilization transform is applied
**Projection matrix** - The original image plane is reprojected to the new image plane and also the coordinates are translated to normalized coordinates. This should probably be separated into two transforms, one projection and one conversion to normalized coordinates, although you often combine these in OpenGL.